Fix handling of non-full clustering keys in the read path

tgrabiec opened the pull request:
This PR fixes several bugs related to handling of non-full
clustering keys.

One is in trim_clustering_row_ranges_to(), which is broken for non-full keys in reverse
mode. It will trim the range to position_in_partition_view::after_key(full_key) instead of
position_in_partition_view::before_key(key), hence it will include the
key in the resulting range rather than exclude it.

Fixes #12180

after_key() was creating a position which is after all keys prefixed
by a non-full key, rather than a position which is right after that
key.

This will issue will be caught by cql_query_test::test_compact_storage
in debug mode when mutation_partition_v2 merging starts inserting
sentinels at position after_key() on preemption.

It probably already causes problems for such keys as after_key() is used 
in various parts in the read path.

Refs #1446

tgrabiec commented:
Extracted from https://github.com/scylladb/scylladb/pull/12048

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3398/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3414/

tgrabiec commented:
In V2:

  - Introduced position_in_partition_view::after_key() which tries to avoid memory allocation when possible: https://github.com/scylladb/scylladb/pull/12234/commits/44ac22c2fe3e28899cd3ffa5e0a174744b4ab270#diff-09a5f264dc2ca607e8435146a2bbc747677c0696c71535c4a2c1103f0a537b74R607
   The current cache reader implementation uses it on the hot path in several places.

- Added patch: types: Fix comparison of frozen sets with empty values

tgrabiec commented:
Ale the cache_flat_mutation_reader.hh code was adjusted to use the new after_key() variant, please review carefully: https://github.com/scylladb/scylladb/pull/12234/commits/44ac22c2fe3e28899cd3ffa5e0a174744b4ab270#diff-6189e1eabce7b94b6ce6182b38f139fe2f916d6250d818faae81d2b8ce6f4700

@michoecho @denesb

tgrabiec commented:
A general comment: I don't like the fact that after_key() needs to reallocate the key. It would be better to introduce a new bound_kind which would order the position where we want it. But this seems like a very risky and effortful change, because there is a bunch of code which inspects the bound weight and works on the assumption that there are only certain kinds of weights. All those would have to be carefully adjusted.

denesb commented on position_in_partition.hh:
```diff
@@ -572,6 +598,25 @@ public:
+    if (pos.key().is_full(s)) {
+        return {position_in_partition_view::after_all_prefixed(pos.key()), std::nullopt};
+    }
+    // FIXME: This wouldn't be needed if we had a bound weight to represent this.
```
I think we should do this, but not in this PR, as we will likely want to backport this fix.
We can easily locate all sites which look at weight explicitely and consider the effect of the new weights for each one. There shouldn't be a lot of them, most should be in this very file.

denesb commented on cache_flat_mutation_reader.hh:
```diff
@@ -832,7 +832,8 @@ inline
     auto& row = mf.as_clustering_row();
-    auto new_lower_bound = position_in_partition::after_key(row.key());
+    auto new_lower_bound = position_in_partition(
+            position_in_partition_view::after_key(*_schema, row.key()).view);
```
Why not use `position_in_partition::after_key()` here?

tgrabiec replied:
Yes, we could.

denesb commented on position_in_partition.hh:
```diff
@@ -308,6 +312,13 @@ public:
+    // If given position is a clustering row position, returns a position
+    // right before it. Otherwise, returns it unchanged.
+    // The position "pos" must be a clustering position.
+    static position_in_partition before_key(position_in_partition_view pos) {
```
Retracting comment. There cannot be key prefixed key values before a prefix key.

tgrabiec commented on query.cc:
```diff
@@ -135,14 +135,13 @@ void trim_clustering_row_ranges_to(const schema& s, clustering_row_ranges& range
     clustering_key::make_full(s, full_key);
-    return trim_clustering_row_ranges_to(s, ranges,
-            reversed ? position_in_partition_view::after_key(full_key) : position_in_partition_view::before_key(full_key), reversed);
+    return trim_clustering_row_ranges_to(s, ranges, position_in_partition_view::before_key(full_key), reversed);
```
In the reversed case, we also need to avoid make_full(), cause before_key(make_full(x)) would be wrong, it would not exclude the key.

tgrabiec commented:
I realized that two more things are currently broken for non-full keys (before the PR):

  - p_i_p::reversed()
  
  - reversed-schema comparator for two keys where one is full and one is non-full

tgrabiec commented:
The comparator will order (keys) {a} before {a, b} for both forward and reverse schemas.

I checked Cassandra, and it behaves the same, that is it orders null component before non-null component even with `WITH CLUSTERING ORDER BY (... DESC);`. So we cannot construct a reversed schema based on column definitions alone, which will correctly order non-full keys.

tgrabiec commented:
Changing clustering order to the reverse during query on the other hand, puts the null after the rows. What a mess.

tgrabiec commented:
@avikivity @bhalevy @tzach I need a decision, are we investing into fixing the mess with non-full keys, or maybe it's better to deprecate and reject them?

tgrabiec commented:
Refs https://github.com/scylladb/scylladb/issues/12258

avikivity commented:
> @mykaul @avikivity @bhalevy @tzach I need a decision, are we investing into fixing the mess with non-full keys, or maybe it's better to deprecate and reject them?

IIRC they can only be created via Thrift, yes?

If so let's deprecate+reject.

Plan:
1. deprecate and eventually remove thrift
2. reject non-full clustering keys in thrift
3. add a tool to detect if a node has sstables with non-full clustering keys (another validate mode? @denesb )
4. if anyone reports positive results from the tool, add another tool to scrub them out. Or maybe just have the readers drop fields with non-full clustering keys.

tgrabiec commented:
On Sun, Dec 11, 2022 at 12:38 PM Avi Kivity ***@***.***>
wrote:

> @mykaul <https://github.com/mykaul> @avikivity
> <https://github.com/avikivity> @bhalevy <https://github.com/bhalevy>
> @tzach <https://github.com/tzach> I need a decision, are we investing
> into fixing the mess with non-full keys, or maybe it's better to deprecate
> and reject them?
>
> IIRC they can only be created via Thrift, yes?
>

No, you can also create them via CQL on tables created "with compact
storage". See the example in the recently linked github issue.

> If so let's deprecate+reject.
>
> Plan:
>
>    1. deprecate and eventually remove thrift
>    2. reject non-full clustering keys in thrift
>    3. add a tool to detect if a node has sstables with non-full
>    clustering keys (another validate mode? @denesb
>    <https://github.com/denesb> )
>    4. if anyone reports positive results from the tool, add another tool
>    to scrub them out. Or maybe just have the readers drop fields with non-full
>    clustering keys.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylladb/pull/12234#issuecomment-1345530143>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACFILYOISE6GCXMVX3DCETWMW4KZANCNFSM6AAAAAASW5VJNM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

avikivity commented:
Well, it's a lot harder to be sure there are no active applications that use partial clustering keys.

tgrabiec commented:
In order to make progress with https://github.com/scylladb/scylladb/pull/12048, I suggest to proceed with the fixes made here for forward reads and ignore #12258 for now. If we decide to drop non-full key support, we can remove the handling added here later.

tgrabiec commented:
In V3:
 - Simplify one usage of after_key() in cache: https://github.com/scylladb/scylladb/compare/44ac22c2fe3e28899cd3ffa5e0a174744b4ab270..23e4c83155332d3f0607f16ee8829ca4cd75d9be#diff-6189e1eabce7b94b6ce6182b38f139fe2f916d6250d818faae81d2b8ce6f4700

 - More comprehensive unit test

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3498/
